### PR TITLE
Adding catch clauses to the CFG, updaging SE engine to support branch…

### PR DESCRIPTION
**Pull request by valhristov**
… blocks that are not created from switch statements

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -479,16 +479,27 @@ private void BuildLabeledStatement(LabeledStatementSyntax labeledStatement)
...
-            currentBlock = CreateBlock(currentBlock);
+            var catchBlocks = new List<Block>();
+            foreach (var catchClause in tryStatement.Catches)
+            {
+                var catchBlock = currentBlock = CreateBlock(finallyBlock);
+                catchBlocks.Add(catchBlock);
+                BuildBlock(catchClause.Block);
+            }
+
+            var tryEndBlock = currentBlock = CreateBranchBlock(tryStatement, catchBlocks.Union(new[] { finallyBlock }));
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: MAJOR Extract the assignment of 'currentBlock' from this expression. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1121)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -479,16 +479,27 @@ private void BuildLabeledStatement(LabeledStatementSyntax labeledStatement)
 
         private void BuildTryStatement(TryStatementSyntax tryStatement)
         {
+            var finallyBlock = currentBlock;
             if (tryStatement.Finally?.Block != null)
             {
-                currentBlock = CreateBlock(currentBlock);
+                finallyBlock = currentBlock = CreateBlock(currentBlock);
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: MAJOR Extract the assignment of 'currentBlock' from this expression. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1121)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -479,16 +479,27 @@ private void BuildLabeledStatement(LabeledStatementSyntax labeledStatement)
...
-                currentBlock = CreateBlock(currentBlock);
+                finallyBlock = currentBlock = CreateBlock(currentBlock);
                 BuildBlock(tryStatement.Finally.Block);
             }
 
-            currentBlock = CreateBlock(currentBlock);
+            var catchBlocks = new List<Block>();
+            foreach (var catchClause in tryStatement.Catches)
+            {
+                var catchBlock = currentBlock = CreateBlock(finallyBlock);
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: MAJOR Extract the assignment of 'currentBlock' from this expression. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1121)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: MAJOR Extract the assignment of 'currentBlock' from this expression. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1121)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs (hunk 2 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: MAJOR Extract the assignment of 'currentBlock' from this expression. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1121)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs (hunk 3 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: MAJOR Extract the assignment of 'currentBlock' from this expression. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1121)

File: sonaranalyzer-dotnet/its/expected/Ember-MM/Ember.Plugins-{9496C697-5AFD-4813-AEDC-AF33FACEADF0}-S2259.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
+"issues":  [
+{
+"id":  "S2259",
+"message":  "'className' is null on at least one execution path.",
+"location":  {
+"uri":  "Ember-MM\Ember.Plugins\PluginSectionHandler.cs",
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/its/expected/Ember-MM/Ember.Plugins-{9496C697-5AFD-4813-AEDC-AF33FACEADF0}-S2259.json]: Great stuff, this alone shows the exercise was worthwhile.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 5]
```diff
@@ -851,6 +851,12 @@ protected override bool IsValueConsumingStatement(SyntaxNode jumpNode)
                 return returnStatement.Expression != null;
             }
 
+            var switchStatement = jumpNode as SwitchStatementSyntax;
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: What is that for?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs [hunk 6]
```diff
@@ -1483,133 +1483,200 @@ public void Cfg_TryFinally()
...
+                cw4();
+            }
+            cw5();");
 
-            VerifyCfg(cfg, 4);
+            VerifyCfg(cfg, 7);
 
             var blocks = cfg.Blocks.ToList();
 
             var tryStartBlock = blocks[0];
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs]: As suggested above - perhaps do that?
var beforeTryBlock = blocks[0];
var tryBlock = blocks[1];

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/ControlFlowGraph/SyntaxNodeBased/ControlFlowGraphBuilder.cs [hunk 7]
```diff
@@ -84,7 +85,7 @@ internal IControlFlowGraph Build()
             {
                 Blocks = Blocks,
                 EntryBlock = EntryBlock,
-                ExitBlock = ExitBlock
+                ExitBlock = (ExitBlock)ExitTarget.Pop()
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/ControlFlowGraph/SyntaxNodeBased/ControlFlowGraphBuilder.cs]: Shall we check here if the stack is empty after pop?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs [hunk 8]
```diff
@@ -1483,133 +1483,200 @@ public void Cfg_TryFinally()
...
 
-            VerifyAllInstructions(blocks[0], "cw0", "cw0()");
-            VerifyAllInstructions(blocks[1], "cw1", "cw1()");
-            VerifyAllInstructions(blocks[2], "cw2", "cw2()");
-            VerifyAllInstructions(blocks[3], "cw3", "cw3()");
-            VerifyAllInstructions(blocks[4], "cw4", "cw4()");
-            VerifyAllInstructions(blocks[5], "cw5", "cw5()");
-            VerifyAllInstructions(blocks[6]);
+            var tryStartBlock = blocks[0];
+            var tryReturnBlock = blocks[1];
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs]: Slightly confusing name. This is not a try return block. It's a try body that happens to have a return instruction.

I'd suggest it to:
var beforeTryBlock = blocks[0];
var tryBlock = blocks[1]; or tryBodyBlock

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs [hunk 9]
```diff
@@ -1483,133 +1483,200 @@ public void Cfg_TryFinally()
...
+
+            tryStartBlock.Should().BeOfType<BranchBlock>();
+            VerifyAllInstructions(tryStartBlock, "cw0", "cw0()");
+            tryStartBlock.SuccessorBlocks.Should().BeEquivalentTo(new[] { tryReturnBlock, catchBlock });
+
+            tryReturnBlock.Should().BeOfType<JumpBlock>();
+            VerifyAllInstructions(tryReturnBlock, "cw1", "cw1()");
+            tryReturnBlock.SuccessorBlocks.Should().BeEquivalentTo(new[] { finallyBlock });
+
+            tryEndBlock.Should().BeOfType<BranchBlock>();
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs]: Shouldn't tryEndBlock have tryReturnBlock as a predecessor?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Verifier.cs [hunk 10]
```diff
@@ -66,13 +64,19 @@ internal static class Verifier
             using (var workspace = new AdhocWorkspace())
             {
                 var file = new FileInfo(path);
-                var project = CreateProject(file, GeneratedAssemblyName, workspace).AddDocument(file);
+                var project = CreateProject(file.Extension, GeneratedAssemblyName, workspace).AddDocument(file);
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Verifier.cs]: What are changes in this file about?

Comment (valhristov) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Verifier.cs]: Supporting code snippets in tests. Very useful for debugging.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs (hunk 5 above)
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: This method determines whether a value should be popped from the stack. Now we are handling both switch and try/catch in VisitBranchBlock in Common.ExplodedGraph, hence we need a way to determine whether to pop (switch) or not to pop (try) from stack...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/ControlFlowGraph/SyntaxNodeBased/ControlFlowGraphBuilder.cs (hunk 7 above)
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/ControlFlowGraph/SyntaxNodeBased/ControlFlowGraphBuilder.cs]: Good suggestion.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs (hunk 9 above)
Comment (valhristov) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs]: I think you may be right - the tryEndBlock seems not accessible in this case. We could probably optimize for blocks that end with return/break/continue/throw. I will try to improve with the next ticket about nested try-catch.

**Pull request merged**